Task 1

In [ ]:
import pandas as pd
import numpy as np

df = pd.read_csv('adult.csv')

print("initial information")
df.info()

print("\nfirst five rows")
print(df.head())

print("\nnull count before fix")
print(df.isnull().sum())


df.replace('?', np.nan, inplace=True)

print("\nnull count after fix")
print(df.isnull().sum())

initial information
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB

first five rows
   age workclass  fnlwgt     education  education.num ma

Task 2

In [ ]:
missingvals = df.isnull().sum()

missingpercent = df.isnull().mean() * 100

report = pd.DataFrame({
    'Missing Count': missingvals,
    'Missing Percentage (%)': missingpercent
})

# sort the values
sortedreport = report.sort_values(by='Missing Count', ascending=False)

print("Missing values and the percentage sorted")
print(sortedreport)

print("\ncolumns with most missing values")
print("occupation, workclass, native.country (in that order)")

Missing values and the percentage sorted
                Missing Count  Missing Percentage (%)
occupation               1843                5.660146
workclass                1836                5.638647
native.country            583                1.790486
fnlwgt                      0                0.000000
education                   0                0.000000
education.num               0                0.000000
age                         0                0.000000
marital.status              0                0.000000
relationship                0                0.000000
sex                         0                0.000000
race                        0                0.000000
capital.gain                0                0.000000
capital.loss                0                0.000000
hours.per.week              0                0.000000
income                      0                0.000000

columns with most missing values
occupation, workclass, native.country (in that order)


Task 3

In [ ]:
# for categorical, fill with unkown
catcols = ['workclass', 'occupation', 'native.country']
for col in catcols:
    if col in df.columns:
        df[col].fillna('Unknown', inplace=True)

# fill numerical with medians
numericcols = df.select_dtypes(include=['int64', 'float64']).columns
for col in numericcols:
    if df[col].isnull().sum() > 0:
        df[col].fillna(df[col].median(), inplace=True)

print("Verify if any missing left:")
print(df.isnull().sum().sum(), "total missing values remaining.")

Verify if any missing left:
0 total missing values remaining.


Task 4

In [ ]:
duplicates = df.duplicated().sum()
print(f"Exact duplicate rows: {duplicates}")

colsnoincome = [col for col in df.columns if col != 'income'] #creating a list of all columns to ignore income (the target col)

# find duplicates now
duplicateswithoutincome = df.duplicated(subset=colsnoincome).sum()
print(f"Duplicates ignoring the 'income' column: {duplicateswithoutincome}")

#drop only exact duplicates
df.drop_duplicates(inplace=True)
print(f"Rows remaining after dropping exact duplicates: {len(df)}")

Exact duplicate rows: 24
Duplicates ignoring the 'income' column: 25
Rows remaining after dropping exact duplicates: 32537


Task 5

In [ ]:
cols_to_clean = ['education', 'marital.status', 'native.country']

print("unique values before cleaning")
for col in cols_to_clean:
    print(f"\n{col} unique values:")
    print(df[col].unique())


for col in cols_to_clean:
    df[col] = df[col].astype(str).str.strip() # astype to ensure that data type is of string

#verification
print("\nafter cleaning")
for col in cols_to_clean:
    print(f"\n{col} unique values:")
    print(df[col].unique())

unique values before cleaning

education unique values:
['HS-grad' 'Some-college' '7th-8th' '10th' 'Doctorate' 'Prof-school'
 'Bachelors' 'Masters' '11th' 'Assoc-acdm' 'Assoc-voc' '1st-4th' '5th-6th'
 '12th' '9th' 'Preschool']

marital.status unique values:
['Widowed' 'Divorced' 'Separated' 'Never-married' 'Married-civ-spouse'
 'Married-spouse-absent' 'Married-AF-spouse']

native.country unique values:
['United-States' 'Unknown' 'Mexico' 'Greece' 'Vietnam' 'China' 'Taiwan'
 'India' 'Philippines' 'Trinadad&Tobago' 'Canada' 'South'
 'Holand-Netherlands' 'Puerto-Rico' 'Poland' 'Iran' 'England' 'Germany'
 'Italy' 'Japan' 'Hong' 'Honduras' 'Cuba' 'Ireland' 'Cambodia' 'Peru'
 'Nicaragua' 'Dominican-Republic' 'Haiti' 'El-Salvador' 'Hungary'
 'Columbia' 'Guatemala' 'Jamaica' 'Ecuador' 'France' 'Yugoslavia'
 'Scotland' 'Portugal' 'Laos' 'Thailand' 'Outlying-US(Guam-USVI-etc)']

after cleaning

education unique values:
['HS-grad' 'Some-college' '7th-8th' '10th' 'Doctorate' 'Prof-school'
 'Bachel

Task 6

In [ ]:
print("summary of numerical columns")
print(df.describe())

print("\ncategorical summary")
print(df.describe(include='object'))

age_mean = df['age'].mean()
age_median = df['age'].median()

hours_mean = df['hours.per.week'].mean()
hours_median = df['hours.per.week'].median()

print(f"Age Mean: {age_mean:.2f}, Age Median: {age_median:.2f}")
print(f"Hours Mean: {hours_mean:.2f}, Hours Median: {hours_median:.2f}")

summary of numerical columns
                age        fnlwgt  education.num  capital.gain  capital.loss  \
count  32537.000000  3.253700e+04   32537.000000  32537.000000  32537.000000   
mean      38.585549  1.897808e+05      10.081815   1078.443741     87.368227   
std       13.637984  1.055565e+05       2.571633   7387.957424    403.101833   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.369930e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours.per.week  
count    32537.000000  
mean        40.440329  
std         12.346889  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000  

categorical summary
       workcl

Task 7

In [ ]:
print("Most Common Occupation")
occupation_counts = df['occupation'].value_counts()
print(occupation_counts.head(1))

print("\nGender Dist.")
genderdist = df['sex'].value_counts(normalize=True) * 100
print(genderdist)

# c) Percentage distribution of the income label
print("\nIncome Dist.")
incomedist = df['income'].value_counts(normalize=True) * 100
print(incomedist)

Most Common Occupation
occupation
Prof-specialty    4136
Name: count, dtype: int64

Gender Dist.
sex
Male      66.92381
Female    33.07619
Name: proportion, dtype: float64

Income Dist.
income
<=50K    75.907428
>50K     24.092572
Name: proportion, dtype: float64


Task 8


In [ ]:
edumap = df.groupby('education')['education.num'].unique()

print("Mapping")
print(edumap)


inconsistencies = edumap[edumap.apply(len) > 1]

print("\nCheck inconsistency")
if len(inconsistencies) == 0:
    print("No inconsistencies in data.")
else:
    print("Inconsistencies found:")
    print(inconsistencies)

Mapping
education
10th             [6]
11th             [7]
12th             [8]
1st-4th          [2]
5th-6th          [3]
7th-8th          [4]
9th              [5]
Assoc-acdm      [12]
Assoc-voc       [11]
Bachelors       [13]
Doctorate       [16]
HS-grad          [9]
Masters         [14]
Preschool        [1]
Prof-school     [15]
Some-college    [10]
Name: education.num, dtype: object

Check inconsistency
No inconsistencies in data.


Task 9

In [ ]:
import sqlite3
import pandas as pd

conn = sqlite3.connect('adult_income_db.sqlite')

raw_df = pd.read_csv('adult.csv')
raw_df.to_sql('adult_income', conn, if_exists='replace', index=False)


sql_query = "SELECT * FROM adult_income WHERE age > 30"

sql_df = pd.read_sql_query(sql_query, conn)

print("Shape compared.")
print(f"Original CSV Shape (All Data): {raw_df.shape}")
print(f"SQL Extracted DataFrame Shape (Age > 30): {sql_df.shape}")

conn.close()

--- Shape Comparison ---
Original CSV Shape (All Data): (32561, 15)
SQL Extracted DataFrame Shape (Age > 30): (21989, 15)


Task 10

# Data Profiling Summary Report

The dataset has 15 columns and 32561 rows prior to cleaning and dropping etc. It is representing data of different adults and their incomes.

Issues with data:
Missing data was denoted by '?' instead of actual NaN values. In reality, there are 1800+ missing values. It also had some duplicates across.

1 to 1 mapping of education was verified.

Cleaning decisions:
Missing values of categorical type were replaced with "unkown" to retain the rows, while numerical ones were replaced with median.
Duplicates that were exact were dropped, but those differing only on income were kept.
Inconsitent data was cleaned up using str.strip, but I didn't observe any significant change.